src/utils/histogram.cc

```diff
@@ -0,0 +1,138 @@
+/*
+ This file is part of Leela Chess Zero.
+ Copyright (C) 2018 The LCZero Authors
+
+ Leela Chess is free software: you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ Leela Chess is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Leela Chess.  If not, see <http://www.gnu.org/licenses/>.
+
+ Additional permission under GNU GPL version 3 section 7
+
+ If you modify this Program, or any covered work, by linking or
+ combining it with NVIDIA Corporation's libraries from the NVIDIA CUDA
+ Toolkit and the the NVIDIA CUDA Deep Neural Network library (or a
+ modified version of those libraries), containing parts covered by the
+ terms of the respective license agreement, the licensors of this
+ Program grant you additional permission to convey the resulting work.
+ */
+
+#include <algorithm>
+#include <cmath>
+#include <iomanip>
+#include <iostream>
+#include <stdio.h>
+#include <string.h>
+#include "utils/histogram.h"
+
+namespace lczero {
+
+namespace {
+void Print(const std::string& what) { std::cerr << what; }
+
+void PrintAligned(const std::string& what, int aligned) {
+  std::cerr << std::right << std::setw(aligned) << what;
+}
+
+std::string Format(const std::string& format, double value) {
+  static const int kMaxBufferSize = 32;
+  char buffer[kMaxBufferSize];
+  int len = snprintf(buffer, kMaxBufferSize, format.c_str(), value);
+  return std::string(buffer, buffer + len);
+}
+} // namespace
+
+Histogram::Histogram()
+    : Histogram(kDefaultMinExp, kDefaultMaxExp, kDefaultMinorScales) {}
+
+Histogram::Histogram(int min_exp, int max_exp, int minor_scales)
+    : min_exp_(min_exp),
+      max_exp_(max_exp),
+      minor_scales_(minor_scales),
+      major_scales_(max_exp_ - min_exp_ + 1),
+      total_scales_(major_scales_ * minor_scales_),
+      full_scales_(total_scales_ + 4),
+      buckets_(full_scales_) {
```

Comment:
Is there a need for `full_scales_` if you can use `buckets_.size()` ?

Reply:
I will make the changes, but it's slightly less readable in my opinion.
"full_scales" is gone.  But it was both a member and a comment.

Reply:
It's fine to keep it, but the it makes sense to add `assert(full_scales_ == buckets_.size())` in Reset()